Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+        // This state is valid: server can send trailing metadata without END_STREAM
+        // set, and follow it with an empty message frame where the flag *is* set.
+        ()
+        // TODO: I believe we should set some flag in the state to signal that
+        // we're expecting an empty data frame with END_STREAM set; otherwise,
+        // we could get an infinite number of metadata frames from the server -
+        // not sure this should be allowed.
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is closed, nothing could have been sent.")
+      case .clientClosedServerClosed, .clientClosedServerIdle, .clientClosedServerOpen:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is closed, shouldn't have received anything.")
+      }
+    }
+    
+    mutating func receive(message: ByteBuffer, endStream: Bool) throws {
+      // This is a message received by the client, from the server.
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Cannot have received anything from server if client is not yet open.")
+      case .clientOpenServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server cannot have sent a message before sending the initial metadata.")
+      case .clientOpenServerOpen(var state):
+        try state.deframer.process(buffer: message) { deframedMessage in
+          state.inboundMessageBuffer.append(deframedMessage)
+        }
+        if endStream {
+          self.state = .clientOpenServerClosed(.init(previousState: state))
+        } else {
+          self.state = .clientOpenServerOpen(state)
+        }
+      case .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Cannot have received anything from a closed server.")
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Shouldn't receive anything if client's closed.")
+      }
+    }
+    
+    mutating func nextInboundMessage() -> [UInt8]? {
+      switch self.state {
+      case .clientOpenServerOpen(var state):
+        let message = state.inboundMessageBuffer.pop()
+        self.state = .clientOpenServerOpen(state)
+        return message
+      case .clientOpenServerClosed(var state):
+        let message = state.inboundMessageBuffer.pop()
+        self.state = .clientOpenServerClosed(state)
+        return message
+      case .clientOpenServerIdle, 
+          .clientIdleServerIdle, 
+          .clientClosedServerIdle,
+          .clientClosedServerOpen,
+          .clientClosedServerClosed:
+        return nil
+      }
+    }
+    
+    private func assertionFailureAndCreateRPCErrorOnFailedPrecondition(_ message: String) -> RPCError {
+      if !self.skipAssertions {
+        assertionFailure(message)
+      }
```

Comment:
Not that it matters much but you could do `assert(self.skipAssertions, message)`